ballista/rust/scheduler/src/state/execution_graph.rs

```diff
@@ -153,6 +157,7 @@ impl ExecutionStage {
             task_statuses: vec![None; num_tasks],
             output_link,
             resolved,
+            stage_metrics: None,
```

Comment:
I'm not sure about storing the raw metrics in the `ExecutionGraph`. It could potentially be a lot of data (eg if you had a `ParquetExec` across many files then you can have a potentially very big set of metrics) and we have to read/write the graph to the state store a lot. 

I think it might be worth introducing a `MetricStore` trait where we can make the storage (or non-storage as the case may be) of metrics extensible.

Reply:
@thinkharderdev Actually the `stage_metrics` stored here is not the raw metrics. It is combined/aggregated among all the tasks of that stage, that's why I name it to `stage_metrics`, not task metrics.

Reply:
Right, that helps but I still think that storing the metrics in the `ExecutionGraph` is not the right way to go. Even aggregated we're adding more data to a value that has to be read/written/decoded/encoded a lot. The other problem I think is that ideally we want to discard the full `ExecutionGraph` soon after the job is completed to prevent unbounded growth in the state store. But metrics may be something that we wish to preserve for longer (or even indefinitely). In addition, I think this may very well be an area where extensibility could be important. The `ExecutionGraph` is an internal implementation detail, but users may have an interest in storing and analyzing metrics and may wish to "export" them to a different system of their choosing (RDBMS, etc). Having an interface that allows for that would be helpful.

Reply:
> 
> Even aggregated we're adding more data to a value that has to be read/written/decoded/encoded a lot

I agree with you that the `ExecutionGraph` is not the perfect place to store metrics. But since there is no in-memory stat store now. And I do not want to add another key space to sled/etcd to store the metrics, so just reuse the existing `ExecutionGraph`.  I think it is not that heavy compared to the existing task_statuses array If we have thousands of tasks.  

```
/// Status of each already scheduled task. If status is None, the partition has not yet been scheduled
 pub(crate) task_statuses: Vec<Option<task_status::Status>>,
```

Ideally those metrics should be stored in memory because currently they are just for displaying purpose.
For long terms preserving, both the task status and metrics should be persisted to other places.

Reply:
I think we should still have different stores: in-memory stat store,  persisted stat store(sled/etcd) and long term persisted store(RDBMS, object store).